url2a908001cd_malictus_klang_tgz-pJ8-malictus_klang_test_PrimitiveTestJ8/src/malictus/klang/chunk/FORMChunk.java

```diff
@@ -0,0 +1,212 @@
+		}
+	}
+	
+	public void addChunkFromArray(byte[] bytes, String name, int position, RandomAccessFile raff) throws IOException {
+		if (!(this.getParentFile() instanceof EditableFileBase)) {
+			throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+		}
+		long newsize = (this.dataEndPosition - this.dataStartPosition) + bytes.length + 8;
+		if (KlangUtil.adjustForPadByte(bytes.length) != bytes.length) {
+			//grow byte array by 1
+			newsize = newsize + 1;
+			byte[] newBytes = new byte[bytes.length + 1];
+			System.arraycopy(bytes, 0, newBytes, 0, bytes.length);
+			bytes = newBytes;
+		}
+		if (newsize < 0) {
+			throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+		}
+		try {
+			//first write new file size
+			Int4ByteUnsignedBE size = new Int4ByteUnsignedBE();
+			size.setValue(newsize);
+			raff.seek(startPosition + 4);
+			size.writeValueToFile(raff);
+			//now warn parent chunks
+			this.chunkIsAboutToChangeSize(newsize, raff);
+			//now make the addition
+			long startingplace = 0;
+			if (position > this.getSubChunks().size() + 2) {
+				position = this.getSubChunks().size() + 1;
+			}
+			if (position == this.getSubChunks().size() + 1) {
+				if (this.getSubChunks().size() == 0) {
+					startingplace = this.endPosition;
+				} else {
+					startingplace = this.getSubChunks().get(position - 2).getEndPosition();
+				}
+			} else {
+				startingplace = this.getSubChunks().get(position - 1).getStartPosition();
+			}
+			KlangUtil.insertIntoFile(bytes, this.getParentFile().getFile(), startingplace);
+			//insert name and size
+			KlangUtil.insertIntoFile(new byte[8], this.getParentFile().getFile(), startingplace);
+			raff.seek(startingplace);
+			FourCC fcc = new FourCC(name);
+			fcc.writeValueToFile(raff);
+			Int4ByteUnsignedBE chunksize = new Int4ByteUnsignedBE();
+			chunksize.setValue(bytes.length);
+			chunksize.writeValueToFile(raff);
+			//now update and reparse
+			((EditableFileBase)this.getParentFile()).reparseFile(raff);
+			this.chunkJustChanged(raff);
+		} catch (BadValueException err) {
+			throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+		}
+	}
+	
+	public void subChunkWillChangeSize(Chunk chunk, long newsize, RandomAccessFile raf) throws IOException {
+		try {
+			long newDsize = newsize - 8;
+			if (newDsize >= 0) {
+				//write size var to chunk itself
+				Primitive subCsize = new Int4ByteUnsignedBE();	
+				subCsize.setValueFromString(Long.toString(newDsize));
+				raf.seek(chunk.getStartPosition() + 4);
+				subCsize.writeValueToFile(raf);
+			}
+			//now write this chunk's size variable
+			Int4ByteUnsignedBE size = new Int4ByteUnsignedBE();
+			long sizevar = chunk.getEndPosition() - chunk.getStartPosition();
+			//sizevar may be negative now
+			sizevar = newsize - sizevar;
+			long cursize = this.getEndPosition() - this.getStartPosition();
+			size.setValue(cursize + sizevar - 8);
+			raf.seek(startPosition + 4);
+			size.writeValueToFile(raf);
+		} catch (BadValueException err) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)